jobrunner/lib/database.py

```diff
@@ -153,37 +164,104 @@ def get_connection():
     filename = config.DATABASE_FILE
     # Looks icky but is documented `threading.local` usage
     cache = CONNECTION_CACHE.__dict__
+    if filename not in cache:
+        raise Exception("Database not initialised")
+    return cache[filename]
+
+
+class MigrationNeeded(Exception):
+    pass
+
+
+def init_db(filename=None, migrations=None):
+    """Ensure db conn is created and initialised properly.
+
+    Will create new tables, or migrate the exisiting ones as needed.
+    """
+    cache = CONNECTION_CACHE.__dict__
+
+    if filename is None:
+        filename = config.DATABASE_FILE
+    if migrations is None:
+        migrations = MIGRATIONS
+
     if filename in cache:
         return cache[filename]
-    else:
-        connection = get_connection_from_file(filename)
-        cache[filename] = connection
-        return connection
 
-
-def get_connection_from_file(filename):
+    # this allows us to use per-test :memory: dbs
     if str(filename).startswith(":memory:"):
-        filename = ":memory:"
+        db_filename = ":memory:"
+        new_db = True
+    elif filename.exists():
+        db_filename = filename
+        new_db = False
     else:
-        filename.parent.mkdir(parents=True, exist_ok=True)
-    conn = sqlite3.connect(filename)
+        db_filename = filename
+        new_db = True
+        filename.parent.mkdir(exist_ok=True, parents=True)
+
+    conn = sqlite3.connect(db_filename)
     # Enable autocommit so changes made outside of a transaction still get
     # persisted to disk. We can use explicit transactions when we need
     # atomicity.
     conn.isolation_level = None
     # Support dict-like access to rows
     conn.row_factory = sqlite3.Row
-    schema_count = list(conn.execute("SELECT COUNT(*) FROM sqlite_master"))[0][0]
-    if schema_count == 0:
+
+    latest_version = max(migrations) if migrations else 0
```

Comment:
You could also use the default kwarg here:

    `max(migrations, default=0)`

if you wanted to drop the conditional.  I'm not sure it adds much value though.

Reply:
Thats more readable, thanks